### DISCUSSION TITLE
(pybricks-dev): ImportError: cannot import name 'BLEPUPConnection' from 'pybricksdev.connections'

### DISCUSSION THREAD

**Discussion by tiagozip** [pybricks/support#722] [Q&A]
When using the pybricks-dev library, on the examples' code, I get the following error:

**1. Jupyter-notebook**

![image](https://user-images.githubusercontent.com/70700766/188602702-3e8d6165-b233-474b-be44-07c96c03ec66.png)

**2. Visual Studio Code**
![image](https://user-images.githubusercontent.com/70700766/188603570-2540b5a6-df64-4ed4-a562-78e3e2be0172.png)

**Comment by dlech** [comment]
The connection code should look more like this:

```python
from pybricksdev.connections import PybricksHub
from pybricksdev.ble import find_device

hub = PybricksHub()
device = await find_device('Pybricks Hub')
await hub.connect(device)
```

**Reply by tiagozip** [reply]
Thank you for the quick reply!
This time, I get the following error:
```
  File "c:\Users\<USERNAME>\Desktop\<PATH>\main.py", line 5
    device = await find_device('Pybricks Hub')
             ^
SyntaxError: 'await' outside function
```
![image](https://user-images.githubusercontent.com/70700766/188719758-181649c7-2d02-43bb-89aa-01f522479e3e.png)

**Reply by tiagozip** [reply]
And the code I found to solve it returns back to the first error

**Reply by dlech** [reply]
In the first screenshot, you are using a Jupyter notebook, it the second, you are not using a notebook. Which do you actually want to do?

Jupyter notebooks allow top-level await but regular Python does not.

**Reply by dlech** [reply]
Regular Python needs and event loop like this:

```python
import asyncio
...

async def main():
    # code here
    await ...

asyncio.run(main())
```

**Reply by tiagozip** [reply]
Thanks for the answer! But the problem is that, even when running with Jupyter, this returns an error. 😢

![Preview of the error](https://user-images.githubusercontent.com/70700766/188812448-9b6870f9-fa20-45e6-8a7f-f1da5048fdb8.png)

**Reply by dlech** [reply]
The `PybricksHub` class was moved. Try:

```python
from pybricksdev.connections.pybricks import PybricksHub
```

**Reply by tiagozip** [reply]
Oh no, this time I get other error
```
Traceback (most recent call last):
  File "c:\Users\<USERNAME>\Desktop\<PATH>\main.py", line 10, in <module>
    asyncio.run(hub.connect(device))
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.3568.0_x64__qbz5n2kfra8p0\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.3568.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 647, in run_until_complete       
    return future.result()
  File "C:\Users\HUAWEI\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\pybricksdev\connections\pybricks.py", line 162, in connect
    logger.info(f"Connecting to {device.name}")
AttributeError: 'function' object has no attribute 'name'
```
Seems like this is an infinite thing... 😢

**Reply by dlech** [reply]
>     asyncio.run(hub.connect(device))

This looks wrong. Can you share the full program?

**Reply by tiagozip** [reply]
Yes, of course! 😄

![image](https://user-images.githubusercontent.com/70700766/189056898-5ef3d814-1d77-4492-9b7b-b4d71b295318.png)

**Reply by dlech** [reply]
Assuming this is not in a Jupyter notebook, the program should look more like this (everything is done inside the event loop):

```python
import asyncio

from pybricksdev.connections.pybricks import PybricksHub
from pybricksdev.ble import find_device

async def main():
    hub = PybricksHub()

    device = await find_device('Pybricks Hub')
    await hub.connect(device)

    # do stuff with connected hub
    ...

asyncio.run(main())
```

**Reply by tiagozip** [reply]
Hi! Thanks for the response. I used this code:
```
import asyncio

from pybricksdev.connections.pybricks import PybricksHub
from pybricksdev.ble import find_device
from pybricks.pupdevices import DCMotor
from pybricks.parameters import Port
from time import sleep

async def main():
    hub = PybricksHub()
    motor_left = DCMotor(Port.A)
    motor_right = DCMotor(Port.B)

    device = await find_device('Pybricks Hub')
    await hub.connect(device)

    motor_left.dc(50)
    motor_right.dc(-50)
    sleep(1000)
    motor_left.stop()
    motor_right.stop()

asyncio.run(main())
```
The hub is on and connected, near the PC, the PC has Bluetooth enabled, but I receive this timeout:
```
Traceback (most recent call last):
  File "c:\Users\<USER>\Desktop\<PATH>.py", line 23, in <module>
    asyncio.run(main())
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.3568.0_x64__qbz5n2kfra8p0\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.3568.0_x64__qbz5n2kfra8p0\lib\asyncio\base_events.py", line 647, in run_until_complete       
    return future.result()
  File "c:\Users\<USER>\Desktop\<PATH>.py", line 14, in main
    device = await find_device('Pybricks Hub')
  File "C:\Users\<USER>\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\pybricksdev\ble\__init__.py", line 62, in find_device
    raise asyncio.TimeoutError
asyncio.exceptions.TimeoutError
```

**Reply by laurensvalk** [reply]
Before we get into the details, can you describe what you would like to achieve?

It seems that you are currently mixing code that would run on the hub and code that should run on the PC.

Could it be done with code.pybricks.com or is the PC/Python interface required in your application?

**Reply by tiagozip** [reply]
I need the python interface, because with it I can install libraries such as BlueDot
